File: Nikhil/python/execution_context.py
from __future__ import annotations

import logging
import sys
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from enum import Enum
from typing import TYPE_CHECKING, Any, Dict, List, Mapping, MutableMapping, Sequence

# --------------------------------------------------------------------------- #
# Logging configuration (equivalent to Java SimpleLogger)
# --------------------------------------------------------------------------- #
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(name)s [%(levelname)s] %(message)s",
    stream=sys.stderr,
)
logger = logging.getLogger(__name__)

# --------------------------------------------------------------------------- #
# Forward references to avoid circular-import issues
# --------------------------------------------------------------------------- #
if TYPE_CHECKING:  # pragma: no cover
    from steps import Step
    from peace import Peace  # for Person, Task  (already ported)
    from peace import Person as _Person
    from peace import Task as _Task


# --------------------------------------------------------------------------- #
# ExecutionContext
# --------------------------------------------------------------------------- #
class ExecutionContext:
    """
    Holds the *mutable* simulation state during a test run and offers
    utilities for logging and task filtering.
    """

    # ------------------------- life-cycle ------------------------------ #
    def __init__(
        self,
        db: "ExecutionContext.DbConnector",
        scenario_id: str,
        run_id: str,
    ) -> None:
        if db is None:
            raise ValueError("db must not be None")
        if scenario_id is None:
            raise ValueError("scenario_id must not be None")
        if run_id is None:
            raise ValueError("run_id must not be None")

        self.db: ExecutionContext.DbConnector = db
        self.scenario_id: str = scenario_id
        self.run_id: str = run_id

        # Dynamic state ------------------------------------------------- #
        self.unassigned_tasks: list["_Task"] | None = None
        self.operator_tasks: list["_Task"] = []
        self.related_persons: dict[str, "_Person"] | None = None

        self.proforma_document: dict[str, str] = {}
        self.sks: dict[str, str] = {}
        self.poa: dict[str, str] = {}

        self.log_entries: list[ExecutionContext.LogEntry] = []

    # ------------------------------------------------------------------ #
    # Database connector (abstract)                                      #
    # ------------------------------------------------------------------ #
    class DbConnector:
        """A pluggable persistence hook used by the simulator."""

        def add_step(self, run_id: str, step: "Step") -> None:  # noqa: D401
            raise NotImplementedError

    # ------------------------------------------------------------------ #
    # Log entries                                                        #
    # ------------------------------------------------------------------ #
    class LogEntryType(str, Enum):
        API_CALL = "API_CALL"
        PAYMENT = "PAYMENT"
        DIARY_ENTRY = "DIARY_ENTRY"
        INTERACTION = "INTERACTION"
        EMAIL = "EMAIL"
        UPLOAD = "UPLOAD"

        def __str__(self) -> str:  # pragma: no cover
            return str(self.value)

    @dataclass
    class LogEntry:
        """Base class for every log entry."""

        type: "ExecutionContext.LogEntryType"

        def __str__(self) -> str:  # pragma: no cover
            return f">>> {self.type} LOGGED"

    # ------------------------- concrete entries ----------------------- #
    @dataclass
    class ApiCallEntry(LogEntry):
        scenario_id: str
        tool_id: str
        args: dict[str, Any] = field(default_factory=dict)

        def __str__(self) -> str:
            args_formatted = " ".join(f'{k}="{v}"' for k, v in self.args.items())
            return (
                f">>> API CALL LOGGED > {self.scenario_id}: {self.tool_id}({args_formatted})"
            )

    @dataclass
    class DiaryEntry(LogEntry):
        task_time_created: str
        task_customer_number: str
        category: str
        message: str

        def __str__(self) -> str:
            return (
                f">>> DIARY ENTRY LOGGED > For task {self.task_customer_number} - "
                f"{self.task_time_created}\n    [{self.category}] {self.message}"
            )

    @dataclass
    class EmailEntry(LogEntry):
        recipient: str
        message: str

        def __str__(self) -> str:
            return (
                f">>> OUTGOING EMAIL LOGGED >  To: {self.recipient}\nContent: {self.message}"
            )

    @dataclass
    class InteractionEntry(LogEntry):
        message: str

        def __str__(self) -> str:
            return f">>> INTERACTION LOGGED > {self.message}"

    @dataclass
    class PaymentEntry(LogEntry):
        amount: str
        message: str

        def __str__(self) -> str:
            return f">>> PAYMENT LOGGED > Amount: {self.amount} -> {self.message}"

    @dataclass
    class UploadEntry(LogEntry):
        customer_number: str
        document_type: str
        content: str

        def __str__(self) -> str:
            return (
                f">>> {self.document_type} FILE UPLOADED > for client "
                f"{self.customer_number} CONTENT -> {self.content}"
            )

    # ------------------------------------------------------------------ #
    # Logging convenience wrappers                                       #
    # ------------------------------------------------------------------ #
    def log(self, entry: "ExecutionContext.LogEntry") -> None:
        if entry is None:
            raise ValueError("entry must not be None")
        self.log_entries.append(entry)
        logger.info("%s", entry)

    # --- overloaded helpers (mirror Java API) ------------------------- #
    def log_api_call(self, scenario_id: str, tool_id: str, args: Mapping[str, Any]) -> None:
        self.log(
            ExecutionContext.ApiCallEntry(
                ExecutionContext.LogEntryType.API_CALL,
                scenario_id,
                tool_id,
                dict(args),
            )
        )

    def log_diary(
        self,
        task_time_created: str,
        task_customer_number: str,
        category: str,
        message: str,
    ) -> None:
        self.log(
            ExecutionContext.DiaryEntry(
                ExecutionContext.LogEntryType.DIARY_ENTRY,
                task_time_created,
                task_customer_number,
                category,
                message,
            )
        )

    def log_interaction(self, message: str) -> None:
        self.log(
            ExecutionContext.InteractionEntry(
                ExecutionContext.LogEntryType.INTERACTION,
                message,
            )
        )

    def log_payment(self, amount: str, message: str) -> None:
        self.log(
            ExecutionContext.PaymentEntry(
                ExecutionContext.LogEntryType.PAYMENT,
                amount,
                message,
            )
        )

    def clear_log(self) -> None:
        self.log_entries.clear()

    # ------------------------------------------------------------------ #
    # Static helpers – task filtering                                    #
    # ------------------------------------------------------------------ #
    @staticmethod
    def filter_tasks(
        tasks: Sequence["_Task"],
        filter_by: str | None = None,
        filter_value: str | None = None,
        customer_number: str | None = None,
    ) -> list["_Task"]:
        """
        Return tasks matching *customer_number* **and optionally** the
        (*filter_by*, *filter_value*) criterion.

        The filter_by string must correspond to the **JSON alias** of a field
        in the Task model (e.g. ``"Step Name"`` rather than ``step_name``).
        """
        if tasks is None:
            raise ValueError("tasks must not be None")

        def _matches(task: "_Task") -> bool:
            # Filter by customer number first (if provided)
            if customer_number and task.customer_number != customer_number:
                return False

            if filter_by and filter_value is not None:
                # Pydantic exposes alias→field mapping via model_fields
                for fld in task.model_fields.values():  # type: ignore[attr-defined]
                    if fld.alias == filter_by:
                        value = getattr(task, fld.name)
                        return str(value) == filter_value
                # If alias not found, nothing matches
                return False

            return True

        return [t for t in tasks if _matches(t)]

File: Nikhil/java/src/main/java/com/infosys/small/pnbc/ExecutionContext.java
/**
 * 
 */
package com.infosys.small.pnbc;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.infosys.small.pnbc.Peace.Person;
import com.infosys.small.pnbc.Peace.Task;
import com.infosys.small.react.ReactAgent.Step;
import com.kjetland.jackson.jsonSchema.annotations.JsonSchemaDescription;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import lombok.ToString;

/**
 * This stores the execution context, that is all scenario data in the
 * simulation layer that can change during execution. In addition, it allows
 * logging information during the execution of a process that can then be used
 * for debug and testing.
 */
@RequiredArgsConstructor
public class ExecutionContext {

	/**
	 * This is a placeholder for integrating with the database and storing steps.
	 */
	public static abstract class DbConnector {

		/**
		 * Saves one execution step for the given run in the Database.
		 * 
		 * @param runId Unique ID for the current run.
		 * @param step  Step to save.
		 */
		public abstract void addStep(String runId, Step step);
	}

	@Getter
	@Setter
	@AllArgsConstructor(access = AccessLevel.PROTECTED)
	@ToString
	public abstract static class LogEntry {

		public enum Type {
			API_CALL, PAYMENT, DIARY_ENTRY, INTERACTION, EMAIL, UPLOAD
		}

		@JsonProperty(required = true)
		@JsonPropertyDescription("Type of entry.")
		public @NonNull Type type;
	}

	@Getter
	@Setter
	@JsonSchemaDescription("This is the schema for log entries with type==API_CALL; these entries represents tool calls the agent performed.")
	public static class ApiCallEntry extends LogEntry {

		@JsonProperty(required = true)
		@JsonPropertyDescription("Simulated scenario (for testing purposes).")
		private @NonNull String scenarioId;

		@JsonProperty(required = true)
		@JsonPropertyDescription("Unique ID of a tool being invoked by the agent.")
		private @NonNull String toolId;

		@JsonProperty(required = true)
		@JsonPropertyDescription("Parameters passed to the tool.")
		private @NonNull Map<String, Object> args;

		public ApiCallEntry(@NonNull String scenarioId, @NonNull String toolId, @NonNull Map<String, Object> args) {
			super(Type.API_CALL);
			this.scenarioId = scenarioId;
			this.toolId = toolId;
			this.args = new HashMap<>(args);
		}

		@Override
		public String toString() {
			StringBuilder sb = new StringBuilder();
			sb.append(">>> API CALL LOGGED > ").append(scenarioId).append(": ").append(toolId).append("(");
			for (Entry<String, Object> e : args.entrySet()) {
				Object value = e.getValue();
				sb.append(e.getKey()).append("=\"").append(value == null ? "null" : value.toString()).append("\" ");
			}
			sb.append(")");

			return sb.toString();
		}
	}

	@Getter
	@Setter
	@JsonSchemaDescription("This is the schema for log entries with type==DIARY_ENTRY; these are notes registering relevant informations or events for auditing purposes.")
	public static class DiaryEntry extends LogEntry {

		@JsonProperty(required = true)
		@JsonPropertyDescription("The time when the task associated to this entry was created; together with customer number this uniquely identifies a task..")
		public @NonNull String taskTimeCreated;

		@JsonProperty(required = true)
		@JsonPropertyDescription("The customer number for the task associated to this entry; together with creation time this uniquely identifies a task.")
		public @NonNull String taskCustomerNumber;

		@JsonProperty(required = true)
		@JsonPropertyDescription("The category associated to the diary entry.")
		public String category;
		public @NonNull String message;

		public DiaryEntry(@NonNull String taskTimeCreated, @NonNull String taskCustomerNumber, @NonNull String category,
				@NonNull String message) {

			super(Type.DIARY_ENTRY);
			this.taskTimeCreated = taskTimeCreated;
			this.taskCustomerNumber = taskCustomerNumber;
			this.category = category;
			this.message = message;
		}

		@Override
		public String toString() {
			StringBuilder sb = new StringBuilder();
			sb.append(">>> DIARY ENTRY LOGGED > For task ").append(taskCustomerNumber).append(" - ")
					.append(taskTimeCreated).append("\n");
			sb.append("    [").append(category).append("] ").append(message);
			return sb.toString();
		}
	}

	@Getter
	@Setter
	@JsonSchemaDescription("This is the schema for log entries with type==EMAL; these register emails sent out to clients or other entities.")
	public static class EmailEntry extends LogEntry {

		@JsonProperty(required = true)
		@JsonPropertyDescription("Recipient of the email; this can be an email address, a person's name, or a unique customer number, among other things.")
		public String recipient;

		@JsonProperty(required = true)
		@JsonPropertyDescription("The message that you need to send.")
		public String message;

		public EmailEntry(@NonNull String customerNumber, @NonNull String message) {

			super(Type.EMAIL);
			this.recipient = customerNumber;
			this.message = message;
		}

		@Override
		public String toString() {
			StringBuilder sb = new StringBuilder();
			sb.append(">>> OUTGOING EMAIL LOGGED >  To: ").append(message).append("\nContent: ").append(message);
			return sb.toString();
		}
	}

	@Getter
	@Setter
	@JsonSchemaDescription("This is the schema for log entries with type==INTERACTION; these register interactions with operations officer (human in the loop) where they are asked for information, checks, etc.)")
	public static class InteractionEntry extends LogEntry {

		@JsonProperty(required = true)
		@JsonPropertyDescription("A query or message sent to the operations officer (human in the loop).")
		public @NonNull String message;

		public InteractionEntry(@NonNull String message) {

			super(Type.INTERACTION);
			this.message = message;
		}

		@Override
		public String toString() {
			StringBuilder sb = new StringBuilder();
			sb.append(">>> INTERACTION LOGGED ").append(message);
			return sb.toString();
		}
	}

	@Getter
	@Setter
	@JsonSchemaDescription("This is the schema for log entries with type==PAYMENT; these entries instruct the Operations Officer to issue a payment.")
	public static class PaymentEntry extends LogEntry {

		@JsonProperty(required = true)
		@JsonPropertyDescription("An amount that the Operations Officer (human in the loop)was requested to pay.")
		public @NonNull String amount;

		@JsonProperty(required = true)
		@JsonPropertyDescription("Payment details for the operations officer (human in the loop).")
		public @NonNull String message;

		public PaymentEntry(@NonNull String amount, @NonNull String message) {

			super(Type.PAYMENT);
			this.amount = amount;
			this.message = message;
		}

		@Override
		public String toString() {
			StringBuilder sb = new StringBuilder();
			sb.append(">>> PAYMENT LOGGED > Amount: ").append(amount).append(" -> ").append(message);
			return sb.toString();
		}
	}

	@Getter
	@Setter
	@JsonSchemaDescription("This is the schema for log entries with type==UPLOAD; these entries logs a file upload in any system.")
	public static class UploadEntry extends LogEntry {

		@JsonProperty(required = true)
		@JsonPropertyDescription("Unique customer number for the estate this dcoument refers to.")
		public @NonNull String customerNumber;

		@JsonProperty(required = true)
		@JsonPropertyDescription("Type of document being uploaded.")
		public @NonNull String documentType;

		@JsonProperty(required = true)
		@JsonPropertyDescription("The content of the document being uploaded.")
		public @NonNull String content;

		public UploadEntry(@NonNull String customerNumber, @NonNull String documentType, @NonNull String content) {
			super(Type.UPLOAD);
			this.customerNumber = customerNumber;
			this.documentType = documentType;
			this.content = content;
		}

		@Override
		public String toString() {
			StringBuilder sb = new StringBuilder();
			sb.append(">>> ").append(documentType).append(" FILE UPLOADED > for client ").append(customerNumber)
					.append(" CONTENT -> ").append(content);
			return sb.toString();
		}
	}

	/**
	 * Connector to the database (for storing execution state).
	 */
	@Getter
	private final DbConnector db;

	/**
	 * Unique ID for the scenario we are running.
	 */
	@Getter
	private final String scenarioId;

	/**
	 * Unique ID for the current run.
	 */
	@Getter
	private final String runId;

	/**
	 * List of unassigned tasks; it contains ALL unassigned tasks (non filtered).
	 */
	@Getter
	@Setter
	private List<Task> unassignedTasks = null;

	/**
	 * List of tasks assigned to operator 42.
	 */
	@Getter
	@Setter
	private List<Task> operatorTasks = new ArrayList<>();

	/** All persons related to the estate, by their Customer Number. */
	@Getter
	@Setter
	private Map<String, Person> relatedPersons = null;

	/** Proforma Document, if provided, by customer number. */
	@Getter
	private final Map<String, String> proformaDocument = new HashMap<>();

	/** SKS (Probate document), if provided, by customer number. */
	@Getter
	private final Map<String, String> SKS = new HashMap<>();

	/** Power of Attorney document, if provided by customer number. */
	@Getter
	private final Map<String, String> PoA = new HashMap<>();

	/** List of log entries. */
	@Getter
	private List<LogEntry> logEntries = new ArrayList<>();

	public void log(@NonNull LogEntry entry) {
		logEntries.add(entry);
		System.err.println(entry.toString());
	}

	/**
	 * Logs an API call.
	 */
	public void log(@NonNull String scenarioId, @NonNull String toolId, @NonNull Map<String, Object> args) {
		log(new ApiCallEntry(scenarioId, toolId, args));
	}

	/**
	 * Logs a diary entry.
	 */
	public void log(@NonNull String taskTimeCreated, @NonNull String taskCustomerNumber, @NonNull String category,
			@NonNull String message) {
		log(new DiaryEntry(taskTimeCreated, taskCustomerNumber, category, message));
	}

	/**
	 * Logs an interaction with the client.
	 */
	public void log(@NonNull String message) {
		log(new InteractionEntry(message));
	}

	/**
	 * Logs a payment.
	 */
	public void log(@NonNull String amount, @NonNull String message) {
		log(new PaymentEntry(amount, message));
	}

	public void clearLog() {
		logEntries.clear();
	}

	public static List<Task> filterTasks(List<Task> tasks, String filterBy, String filterValue, String customerNumber) {
		Stream<Task> stream = tasks.stream();

		// Always filter by customerNumber if provided
		if (customerNumber != null && !customerNumber.isEmpty()) {
			stream = stream.filter(task -> customerNumber.equals(task.customerNumber));
		}

		// Optionally filter by any other column (by JSON name)
		if (filterBy != null && !filterBy.isEmpty() && filterValue != null && !filterValue.isEmpty()) {

			// Find the field in Task with matching @JsonProperty value
			Field matchedField = Arrays.stream(Task.class.getDeclaredFields()).filter(f -> {
				JsonProperty ann = f.getAnnotation(JsonProperty.class);
				return ann != null && filterBy.equals(ann.value());
			}).findFirst().orElse(null);

			if (matchedField != null) {
				matchedField.setAccessible(true);
				stream = stream.filter(task -> {
					try {
						Object value = matchedField.get(task);
						return filterValue.equals(value != null ? value.toString() : null);
					} catch (IllegalAccessException e) {
						// Ignore and skip this filter if reflection fails
						return false;
					}
				});
			}
		}

		return stream.collect(Collectors.toList());
	}
}
